### ISSUE TITLE
MDC not working in Apache Sling

### ISSUE THREAD

**Issue by bryantcj52** [elastic/ecs-logging-java#128]
The automatic MDC keys doesn't happen early enough for them to show up in the logs when using Apache Sling. Error.id was the only one showing up, although not every time for some reason. I had to create a custom Sling Request Filter and manually add the APM trace/transaction/span IDs to MDC.

I'm not sure if it's possible to add this here if it detects the app is running in sling.

https://gist.github.com/bryantcj52/8e5cce3d1a98d3da89b089bcf230ba61

**Comment by eyalkoren** [comment]
Thanks for reporting!

> The automatic MDC keys doesn't happen early enough for them to show up in the logs when using Apache Sling

I am not sure how this would happen. Once we activate a transaction, the IDs are synchronously added to the MDC before the request is being handled. So, theoretically, when your filter is applied, it is already after the agent did just the same (put IDs in MDC) at the beginning of the `FilterChain#doFilter()` invocation as part of the transaction activation.
Can you debug [`MdcActivationListener#before()`](https://github.com/elastic/apm-agent-java/blob/d7de9d2298c71c9dc5bc3ff55001b4a72f40f7fd/apm-agent-plugins/apm-log-correlation-plugin/src/main/java/co/elastic/apm/agent/mdc/MdcActivationListener.java#L186) to see what's not working as expected?

I am assuming Sling already uses slf4j by default. If not, and your logging is done directly through log4j2, note that [version 1.23.0](https://www.elastic.co/guide/en/apm/agent/java/current/release-notes-1.x.html#release-notes-1.23.0) contains a fix to a bug that prevented log correlation when using log4j2 directly.

**Comment by bryantcj52** [comment]
Yes, Sling uses slf4j with logback (https://sling.apache.org/documentation/development/logging.html). I'm on the latest APM agent 1.23.0.

I set up a debug logger for co.elastic.apm.agent.mdc and even co.elastic.apm.agent but nothing is getting logged for some reason. Everything is loaded via osgi so the javaagent may be hooking in before the logging configuration? I tried adding ` -Dorg.slf4j.simpleLogger.defaultLogLevel=debug` to the JVM options but nothing.

**Comment by eyalkoren** [comment]
> Everything is loaded via osgi so the javaagent may be hooking in before the logging configuration?

Good direction! OSGi may be related, not so much in terms of timing, but more so in terms of visibility. 
Can you set `log_level = debug` and share the log (from startup) through https://gist.github.com/?
